Mathlib/Algebra/Module/Submodule/IterateMapComap.lean

```diff
@@ -0,0 +1,110 @@
+/-
+Copyright (c) 2024 Jz Pan. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Jz Pan
+-/
+import Mathlib.Algebra.Module.Submodule.Ker
+
+/-!
+
+# Iterate maps and comaps of submodules
+
+Some preliminary work for establishing the strong rank condition for noetherian rings.
+
+Given two linear maps `f i : M →ₗ[R] M₂` and a submodule `K : Submodule R M`, we can define
+`LinearMap.iterateMapComapAux f i n K : Submodule R M` to be `f⁻¹(i(⋯(f⁻¹(i(K)))))` (`n` times).
+If `f(K) ≤ i(K)`, then this sequence is non-decreasing (`LinearMap.iterateMapComap`).
+If moreover `f` is surjective, `i` is injective, there exists some `m` such that
+`LinearMap.iterateMapComapAux f i m K = LinearMap.iterateMapComapAux f i (m + 1) K`,
+then for any `n`,
+`LinearMap.iterateMapComapAux f i n K = LinearMap.iterateMapComapAux f i (n + 1) K`.
+In particular, by taking `n = 0`, the kernel of `f` is contained in `K`
+(`LinearMap.ker_le_of_iterateMapComapAux_eq_succ`),
+which is a consequence of `LinearMap.ker_le_comap`.
+As a special case, if one can take `K` to be zero,
+then `f` is injective. This is the key result for establishing the strong rank condition
+for noetherian rings.
+
+The construction here is adapted from the proof in Djoković's paper
+*Epimorphisms of modules which must be isomorphisms* [djokovic1973].
+
+-/
+
+open Function Submodule
+
+namespace LinearMap
+
+variable {R M M₂ : Type*} [Semiring R] [AddCommMonoid M] [Module R M]
+  [AddCommMonoid M₂] [Module R M₂] (f i : M →ₗ[R] M₂)
+
+/-- The `LinearMap.iterateMapComapAux f i n K : Submodule R M` is
+`f⁻¹(i(⋯(f⁻¹(i(K)))))` (`n` times). -/
+def iterateMapComapAux (n : ℕ) := (fun K : Submodule R M ↦ (K.map i).comap f)^[n]
```

Comment:
This deserves discussion, but I would call this `iterateMapComap`and find something else for the latter `iterateMapComap`.

Reply:
This naming convention come after the original `tunnelAux` and `tunnel`, one is the map itself, another one is the order preserving map (`→o`).

> ... the latter `iterateMapComap` (in fact, it seems to be never used…)

It is used as `monotone_stabilizes_iff_noetherian.2 ‹_› (f.iterateMapComap i ⊥ (by simp))`, since it is order preserving (i.e. non-decreasing), by Noetherian it must stabilizes.

Reply:
OK, I removed the `→o` version (inlined it in the proof).